File: CsPlugin/Content/Scripts/Cs/Library/Library_Array.js
/// <reference path="../../Cs/Library/Library_Common.ts">/>
// ; typing info for auto-completion in Visual Studio Code

"use strict"

// Library
var NJsCommon = require('Cs/Library/Library_Common.js');

// "typedefs" - classes
/** @type {CommonLibrary} */
var CommonLibrary = NJsCommon.FLibrary;

// "typedefs" - functions
var checkf = CommonLibrary.checkf;
var check = CommonLibrary.check;

module.exports = class NJsArray
{
    static FLibrary = class Library
    {
        /**
         * @param {Array} a 
         * @returns {boolean}
         */
        static /*bool*/ IsEmpty(a /*Array*/)
        {
            if (!Array.isArray(a))
                return true;
            return a.length === 0;
        }

        /**
         * @param {string} context 
         * @param {Array} a 
         * @returns {boolean}
         */
        static /*bool*/ IsNotEmptyChecked(context /*string*/, a /*Array*/)
        {
            // Check is array
            checkf(Array.isArray(a), context + ": a is type: " + typeof a + " and NOT an array.");
            // Check array.length is NOT 0
            checkf(a.length !== 0, context + ": a is EMPTY");
            return true;
        }

        /**
         * @param {Array} a 
         * @returns {boolean}
         */
        static /*bool*/ IsNotEmpty(a /*Array*/)
        {
            // Check is array
            if (!Array.isArray(a))
                return false;
            // Check array.length is NOT 0
            if (a.length === 0)
                return false;
            return true;
        }

        /**
         * @param {Array}   a 
         * @param {number}  size    int
         * @returns {boolean} 
         */
        /*bool*/ static IsSize(a /*Array*/, size /*int*/)
        {
            let self = NJsArray.FLibrary;

            if (!self.IsNotEmpty(a))
                return false;

            if (!CommonLibrary.IsInt(size))
                return false;

            return a.length === size;
        }

        /**
         * @param {string}  context
         * @param {Array}   a 
         * @param {number}  size    int
         * @returns {boolean} 
         */
        /*bool*/ static IsSizeChecked(context /*string*/, a /*Array*/, size /*int*/)
        {
            let self = NJsArray.FLibrary;

            check(self.IsNotEmptyChecked(context, a));

            check(CommonLibrary.IsIntChecked(context, size));

            checkf(a.length === size, context + ": a is NOT of length: " + size);

            return true;
        }

        static AreAllElements_Class_Checked(context, a, classType)
        {
            let self = NJsArray.FLibrary;

            // Check is array
            checkf(Array.isArray(a), context + ": a is type: " + typeof a + " and NOT an array.");

            for (let item of a)
            {
                check(CommonLibrary.IsClassOfChecked(context, item, classType));
            }
            return true;
        }

        static AreAllElements_Class(a, classType)
        {
            // Check is array
            if (!Array.isArray(a))
                return false;

            for (let item of a)
            {
                if (!CommonLibrary.IsClassOf(item, classType))
                    return false;
            }
            return true;
        }

        static AreAllElements2d_Class_Checked(context, a, classType)
        {
            // Check is array
            checkf(Array.isArray(a), context + ": a is type: " + typeof a + " and NOT an array.");

            for (let a2 of a)
            {
                checkf(Array.isArray(a2), context + ": a2 is type: " + typeof a + " and NOT an array.");

                for (let item of a2)
                {
                    check(CommonLibrary.IsClassOfChecked(context, item, classType));
                }
            }
            return true;
        }

        static AreAllElements2d_Class(a, classType)
        {
            // Check is array
            checkf(Array.isArray(a), context + ": a is type: " + typeof a + " and NOT an array.");

            for (let a2 of a)
            {
                checkf(Array.isArray(a2), context + ": a2 is type: " + typeof a + " and NOT an array.");

                for (let item of a2)
                {
                    check(CommonLibrary.IsClassOfChecked(context, item, classType));
                }
            }
            return true;
        }

        static IsNotEmptyAndAllElements_Class_Checked(context, a, classType)
        {
            let self = NJsArray.FLibrary;

            check(self.IsNotEmptyChecked(context, a));

            for (let item of a)
            {
                check(CommonLibrary.IsClassOfChecked(context, item, classType));
            }
            return true;
        }

        static IsNotEmptyAndAllElements_Class(a, classType)
        {
            let self = NJsArray.FLibrary;

            if (self.IsEmpty(a))
                return false;

            for (let item of a)
            {
                if (!CommonLibrary.IsClassOf(item, classType))
                    return false;
            }
            return true;
        }

        static IsNotEmptyAndAllElements2d_Class_Checked(context, a, classType)
        {
            let self = NJsArray.FLibrary;

            check(self.IsNotEmptyChecked(context, a));

            for (let a2 of a)
            {
                check(self.IsNotEmptyChecked(context, a2));

                for (let item of a2)
                {
                    check(CommonLibrary.IsClassOfChecked(context, item, classType));
                }
            }
            return true;
        }

        static IsNotEmptyAndAllElements2d_Class(a, classType)
        {
            let self = NJsArray.FLibrary;

            if (self.IsEmpty(a))
                return false;

            for (let a2 of a)
            {
                if (self.IsEmpty(a2))
                    return false;

                for (let item of a2)
                {
                    if (!CommonLibrary.IsClassOf(item, classType))
                        return false;
                }
            }
            return true;
        }
    };
};

File: CsPlugin/Content/Python/Cs/Library/Library_Array.py
# Copyright 2017-2024 Closed Sum Games, LLC. All Rights Reserved.
# MIT License: https://opensource.org/license/mit/
# Free for use and distribution: https://github.com/closedsum/core

# Library
# - Cs/Library/Library_Common.py
import Cs.Library.Library_Common as Cs_Library_Common

# "typedefs" - library (py)
CommonLibrary = Cs_Library_Common.NPyCommon.FLibrary

# "typedefs" - functions
checkf = CommonLibrary.checkf
check  = CommonLibrary.check

class NPyArray:
    class FLibrary:
        @staticmethod
        def IsEmpty(a: list) -> bool:
            """
            Get whether the Array 'a' is empty or not.

            :param list a:
            :return: bool
            """
            if type(a) is list:
                return len(a) == 0
            return False
        @staticmethod
        def IsNotEmptyChecked(context: str, a: list) -> bool:
            """
            Get whether the Array 'a' is NOT empty or not.
             Assert if NOT.

            :param str context:
            :param list a:
            :return: bool
            """
            # Check is array
            checkf(type(a) is list, context + ": a is type: " + str(type(a)) + " and NOT an array.")
            # Check array.length is NOT 0
            checkf(len(a) != 0, context + ": a is EMPTY")
            return True
        @staticmethod
        def IsNotEmpty(a: list) -> bool:
            """
            Get whether the Array 'a' is NOT empty or not.

            :param list a:
            :return: bool
            """
            # Check is array
            if type(a) is not list:
                return False
            # Check array.length is NOT 0
            if len(a) == 0:
                return False
            return True
        @staticmethod
        def AllNotNoneChecked(context: str, a: list) -> bool:
            """
            Get whether any element in Array 'a' is None or not.
             Assert if NOT.
            
            :param str context;
            :param list a:
            :return: bool
            """
            # Check is array
            checkf(type(a) is list, context + ": a is type: " + str(type(a))+ " and NOT an array.")
            count: int = len(a)
            for i in range(0, count):
                checkf(a[i] != None, context + ": a[" + str(i) + "] is None")
            return True
        @staticmethod
        def IsSize(a: list, size: int) -> bool:
            if type(a) is not list:
                return False

            if type(size) is not int:
                return False
            return len(a) == size
        @staticmethod
        def IsSizeChecked(context: str, a: list, size: int) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            lib.IsNotEmptyChecked(context, a)
            CommonLibrary.IsIntChecked(context, size)
            checkf(len(a) == size, context + ": a is NOT of length: " + size)

            return True
        @staticmethod
        def AreAllElements_Class_Checked(context: str, a: list, classType: type) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            # Check is array
            checkf(type(a) is list, context + ": a is type: " + type(a) + " and NOT an array.")

            for item in a:
                check(CommonLibrary.IsInstanceOfChecked(context, item, classType))
            return True
        @staticmethod
        def AreAllElements_Class(a: list, classType: type) -> bool:
            # Check is array
            if type(a) is not list:
                return False

            for item in a:
                if isinstance(classType, type) == False or isinstance(item, classType) == False:
                    return False
            return True
        @staticmethod
        def Is2dChecked(context: str, a: list) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            checkf(type(a) is list, context + ": a is type: " + type(a) + " and NOT an array.")
            lib.IsNotEmptyChecked(context, a)
            checkf(type(a[0]) is list, context + ": a is NOT a 2d array.")
            return True
        @staticmethod
        def Is2d(a: list) -> bool:
            if type(a) is not list:
                return False
            if len(a) == 0:
                return False
            if type(a[0]) is not list:
                return False
            return True
        @staticmethod
        def AreAllElements2d_Class_Checked(context: str, a: list, classType: type) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            lib.Is2dChecked(context, a)

            for a2 in a:
                for item in a2:
                    check(CommonLibrary.IsInstanceOfChecked(context, item, classType))
            return True
        @staticmethod
        def AreAllElements2d_Class(a: list, classType: type) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            if isinstance(classType, type) == False:
                return False
            if lib.Is2d(a) == False:
                return False

            for a2 in a:
                for item in a2:
                    if CommonLibrary.IsInstanceOf(item, classType) == False:
                        return False
            return True
        @staticmethod
        def IsNotEmptyAndAllElements_Class_Checked(context: str, a: list, classType: type) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            check(lib.IsNotEmptyChecked(context, a))

            for item in a:
                check(CommonLibrary.IsInstanceOfChecked(context, item, classType))
            return True
        @staticmethod
        def IsNotEmptyAndAllElements_Class(a: list, classType: type) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            if type(a) is not list:
                return False
            if len(a) == 0:
                return False
            if isinstance(classType, type) == False:
                return False

            for item in a:
                if isinstance(item, classType) == False:
                    return False
            return True
        @staticmethod
        def IsNotEmptyAndAllElements2d_Class_Checked(context: str, a: list, classType: any) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            check(lib.Is2dChecked(context, a))

            for a2 in a:
                for item in a2:
                    check(CommonLibrary.IsInstanceOfChecked(context, item, classType))
            return True
        @staticmethod
        def IsNotEmptyAndAllElements2d_Class(a: list, classType: type) -> bool:
            lib: NPyArray.FLibrary = NPyArray.FLibrary

            if lib.Is2d(a) == False:
                return False

            for a2 in a:
                for item in a2:
                    if isinstance(item, classType) == False:
                        return False
            return True